### ISSUE TITLE
working with selected amplitudes in CCSD

### ISSUE THREAD

**Issue by SoulokeSen** [pyscf/pyscf#1840]
Dear PYSCF developers,

Is it possible for a given set of orbitals to restrict the single and double amplitudes to certain excitations while solving the amplitues and energy in CCSD ?

thanks and best,
Souloke

**Comment by tberkel** [comment]
No, sorry, I don't think this is possible. It would be a welcome
contribution, but the implementation is nontrivial, especially in a
python/numpy-friendly way.

**Comment by verena-neufeld** [comment]
You can freeze specific orbitals for your ccsd calculation. Not sure whether this helps you.

**Comment by SoulokeSen** [comment]
@tberkel OK, it would be nice to have this feature indeed (especially when working with localized orbitals). I will try to get around the code and see how to best implement this. Could you please refer me to any reference (if there is any specificc implementation) which PYSCF uses for computing the T1 and T2 amplitudes ?
@verena-neufeld  this i already do, but wanted a bit more flexibility to also constrain, in addition, the excitation space.

**Comment by verena-neufeld** [comment]
Would it help to set appropriate entries of t1 and t2 to zero before the end of the update loop? That might be easiest. Depending on how spin sensitive this is, maybe starting with the spinorbital code in gccsd.py is best.

**Comment by tberkel** [comment]
Yes, localized orbitals and the associated sparsity is the natural use case for such a feature. The issue is that pyscf stores the T1 and T2 amplitudes as dense numpy arrays, so that it can do einsum's for fast tensor contraction. If you tried to use explicit for loops and if statements to restrict the excitations, then it would be unbearably slow. One possibility could be to use sparse arrays, but I'm not sure.

If all you want to do is *test* the inclusion/exclusion of certain excitations, then you can indeed zero them out as @verena-neufeld suggested. But you'll of course see no computational savings, you'll just spend time multiplying a lot of zeros ...

**Comment by SoulokeSen** [comment]
@tberkel thanks for your suggestions. I will first try to zero out the necessary amplitudes. But from the point of  computational saving indeed its pointless. One way could be to use slices of the eris, but not sure if that will work

**Comment by verena-neufeld** [comment]
Not saying it’s impossible, just trying to follow: How would you use slices if you say freezing orbitals doesn’t help you and you need to preserve symmetries?

**Comment by SoulokeSen** [comment]
maybe you are right. I might be too premature, but then i guess @tberkel could be a way to use these sparse arrays. But need to think a little
